### PR TITLE
askeladd: EMA decay 0.999→0.9995 — probe unexplored midpoint

### PR DESCRIPTION
**Pull request by morganmcg1**
## Hypothesis

EMA decay 0.999→0.9995: probe the unexplored midpoint between SOTA and known-bad 0.9999.

The current SOTA (PR #115, test_abupt=10.580) uses `--ema-decay 0.999`, giving an effective averaging window of ~1000 steps. We know:
- `ema_decay=0.9999` → ~10,000-step window → confirmed too slow, regressed in earlier rounds
- `ema_decay=0.999` → ~1000-step window → **SOTA**
- `ema_decay=0.998` → ~500-step window → tested in earlier rounds, regressed

The **0.9995 midpoint** (effective window ~2000 steps) has never been tested. It may smooth late-epoch gradient noise and tighten the current val→test generalization gap (1.115×) without the catastrophic slowdown that made 0.9999 fail. This is a clean single-delta experiment.

## Experiment Instructions

Make the following single flag change from SOTA:

```bash
cd target/
torchrun --standalone --nproc_per_node=8 train.py \
  --agent askeladd \
  --optimizer lion --lr 1e-4 --weight-decay 5e-4 \
  --lion-beta1 0.9 --lion-beta2 0.99 \
  --no-compile-model --batch-size 4 --validation-every 1 \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 --model-slices 128 \
  --ema-decay 0.9995 \
  --wandb-group tay-round12-ema-decay-0p9995
```

**Single change from SOTA:** `--ema-decay 0.999` → `--ema-decay 0.9995`

## Evaluation

After training completes, report:
- Best validation `val_primary/abupt_axis_mean_rel_l2_pct` (epoch it occurred)
- Full val trajectory (all 9 epochs)
- All 7 test metrics from summary (list them all, see keys below)
- W&B run ID and URL
- val→test ratio (test / val at best val checkpoint)

**Test metric keys to report:**
- `test_primary/abupt_axis_mean_rel_l2_pct`
- `test_primary/surface_pressure_rel_l2_pct`
- `test_primary/wall_shear_rel_l2_pct`
- `test_primary/volume_pressure_rel_l2_pct`
- `test_primary/wall_shear_x_rel_l2_pct`
- `test_primary/wall_shear_y_rel_l2_pct`
- `test_primary/wall_shear_z_rel_l2_pct`

## Baseline to beat (PR #115 thorfinn)

| Metric | SOTA (PR #115) | AB-UPT |
|---|---:|---:|
| `abupt` mean | **10.580** | — |
| `surface_pressure` | **5.690** | 3.82 |
| `wall_shear` | **10.419** | 7.29 |
| `volume_pressure` | 12.740 | 6.08 |
| `tau_x` | **8.908** | 5.35 |
| `tau_y` | **12.491** | 3.65 |
| `tau_z` | **13.071** | 3.63 |

Best val: 9.484 (ep9) · val→test ratio: 1.115 · W&B run: `d03oghpp`

**Primary metric to beat: `test_primary/abupt_axis_mean_rel_l2_pct` < 10.580 (lower is better)**

### CODE REVIEW

**Comment by morganmcg1** [comment]
STUDENT askeladd:

## Results

**Negative result.** EMA decay 0.9995 regressed vs SOTA (PR #115) on every test metric. The best val was 10.568 at the final epoch (ep9), against SOTA's 9.484, and the test gap to SOTA is +1.04 abupt (+9.8%).

### Summary

| Metric | SOTA (PR #115) | This run (EMA 0.9995) | Δ vs SOTA | AB-UPT |
|---|---:|---:|---:|---:|
| `test_primary/abupt_axis_mean_rel_l2_pct` | **10.580** | 11.619 | **+1.039 (+9.8%)** | — |
| `test_primary/surface_pressure_rel_l2_pct` | **5.690** | 6.465 | +0.775 (+13.6%) | 3.82 |
| `test_primary/wall_shear_rel_l2_pct` | **10.419** | 11.606 | +1.187 (+11.4%) | 7.29 |
| `test_primary/volume_pressure_rel_l2_pct` | **12.740** | 13.195 | +0.455 (+3.6%) | 6.08 |
| `test_primary/wall_shear_x_rel_l2_pct` | **8.908** | 9.883 | +0.975 (+10.9%) | 5.35 |
| `test_primary/wall_shear_y_rel_l2_pct` | **12.491** | 14.035 | +1.544 (+12.4%) | 3.65 |
| `test_primary/wall_shear_z_rel_l2_pct` | **13.071** | 14.514 | +1.443 (+11.0%) | 3.63 |

**Best val:** 10.568 at epoch 9 (last epoch) · SOTA was 9.484 at ep9
**val→test ratio:** 11.619 / 10.568 = **1.099** (vs SOTA 1.115 → tightened by 1.4%)

### Full validation trajectory (`val_primary/abupt_axis_mean_rel_l2_pct`)

| Epoch | step | abupt | surface_p | wall_shear | volume_p |
|---:|---:|---:|---:|---:|---:|
| 1 | 2720 | 77.071 | 63.669 | 83.282 | 60.124 |
| 2 | 5441 | 41.018 | 30.646 | 45.249 | 23.733 |
| 3 | 8162 | 21.702 | 14.951 | 23.902 | 13.805 |
| 4 | 10883 | 15.048 | 10.101 | 16.533 | 9.801 |
| 5 | 13604 | 12.623 | 8.206 | 13.752 | 8.911 |
| 6 | 16325 | 11.151 | 7.200 | 12.355 | 7.206 |
| 7 | 19046 | 10.661 | 6.915 | 11.878 | 6.670 |
| 8 | 21767 | 12.108 | 7.945 | 13.637 | 7.048 |
| 9 | 23526 | **10.568** | 6.832 | 11.842 | 6.393 |

Compared to SOTA's val trajectory ending at 9.484 (ep9), this run is consistently behind by ~1 point — the curve has the same shape, just shifted up because the slower EMA tracks the underlying model improvements with more lag.

### Run details

- **W&B run:** `qvoj4lbs` (rank 0) — group `tay-round12-ema-decay-0p9995`
- **W&B URL:** https://wandb.ai/wandb-applied-ai-team/senpai-v1-drivaerml-ddp8/runs/qvoj4lbs
- **Runtime:** 270.75 min (9 val epochs, hit `SENPAI_TIMEOUT_MINUTES`)
- **Peak memory:** not explicitly recorded in this run (same bf16/batch_size=4/65k-points setup as SOTA, so identical envelope)

### Exact command used

```bash
cd target/
torchrun --standalone --nproc_per_node=8 train.py \
  --agent askeladd \
  --optimizer lion --lr 1e-4 --weight-decay 5e-4 \
  --lion-beta1 0.9 --lion-beta2 0.99 \
  --no-compile-model --batch-size 4 --validation-every 1 \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 --model-slices 128 \
  --ema-decay 0.9995 \
  --wandb-group tay-round12-ema-decay-0p9995
```

### What happened

The hypothesis was that a 0.9995 midpoint between SOTA's 0.999 and known-bad 0.9999 would smooth late-epoch noise and tighten the val→test gap without the catastrophic slowdown of 0.9999. Two observations:

1. **The slowdown is real and dominates.** EMA averaging window is roughly `1/(1-decay)` steps: 0.999 → 1k, 0.9995 → 2k, 0.9999 → 10k. Doubling the window halves the EMA's effective learning rate; in a 9-epoch budget where SOTA's val curve was *still descending at the last epoch*, the slower EMA simply cannot catch up. Best val landed 1.08 abupt above SOTA. The validation trajectory confirms this — same shape, shifted up by ~1 point at every epoch. There was also a transient bump at ep8 (12.108) before recovering at ep9 (10.568), but that's noise, not the central signal.
2. **The val→test gap *did* tighten as predicted, but not enough to matter.** Ratio 1.099 vs SOTA's 1.115 (−1.4%). So the "EMA smooths and improves generalization" story is directionally correct, but the absolute regression on val swamps the gain. With a longer training budget (~15+ epochs) where the slower EMA could fully track the model, this might flip — but at 9 epochs it's a clear loss.

The hypothesis is *empirically* falsified at this training budget. EMA decay scales exponentially with `1-decay`; "midpoint" in decay-space (0.9995) is *not* a midpoint in averaging-window-space (~2000 vs SOTA's ~1000). Treating EMA decay as a linear hyperparameter is misleading.

### Suggested follow-ups

1. **Longer-budget rerun of 0.9995 (~15-18 epochs).** SOTA's best val at ep9 was still descending; if we extend training, 0.9995's slower EMA might finally track the model and the smoother val→test ratio (already 1.099) could pay off. This is the cleanest test of whether EMA 0.9995 is *fundamentally* worse or just *budget-starved*.
2. **Try EMA decay 0.9993 or 0.9994** (effective windows ~1400-1700 steps). If the goal is to tighten val→test without slowing convergence too much, the geometric midpoint between 0.999 (window=1000) and 0.9999 (window=10k) in *log-window space* is ~0.9968 — but in practice 0.9993 is probably the safer next probe.
3. **Stack EMA 0.999 with `--volume-loss-weight 2.0`.** The current SOTA run did NOT include volume_loss_weight=2.0 (per BASELINE.md note); the previous best on volume was PR #111 at 12.548, vs current 12.740. Re-stacking may push abupt below 10.3 as predicted in BASELINE.md without touching EMA.
4. **EMA warmup / cosine-annealed decay.** Start at 0.999 then ramp to 0.9995 in late training — get fast convergence early and smooth EMA late. This could give the best of both regimes within the same 9-epoch budget.

**Comment by morganmcg1** [comment]
ADVISOR: Your W&B run `qvoj4lbs` (group: `tay-round12-ema-decay-0p9995`) has finished. Here are the results I'm reading directly from W&B:

**Final val metrics (ep9):**
- val_abupt: **10.568%** (bar: 9.484 — regression)
- surface_pressure: 6.832%
- vol_pressure: 6.393%
- wall_shear: 11.842%

**Test metrics:**
- test_abupt: **11.619%** (SOTA: 10.580 — regression of +1.04pp)
- test_surface_pressure: 6.465%
- test_vol_pressure: 13.195%
- test_wall_shear: 11.606%
- test_tau_x: 9.883%
- test_tau_y: 14.035%
- test_tau_z: 14.514%

**Analysis:** EMA=0.9995 is clearly worse than SOTA EMA=0.999. All metrics regressed significantly. The slower EMA window (~2000 steps vs ~1000) appears to over-smooth the model weights and hurt generalization. This confirms that 0.999 is the optimal EMA decay for this model/dataset — the 0.998→0.999→0.9995→0.9999 sweep is now complete and 0.999 is the clear winner.

Please post your formal results in a comment and mark the PR ready for review.

**Comment by morganmcg1** [comment]
ADVISOR: Closing PR #194 because Closing as negative result. EMA=0.9995 regressed across all metrics vs SOTA EMA=0.999:
- test_abupt: 11.619% vs SOTA 10.580% (+9.8%)
- val_abupt: 10.568% vs SOTA 9.484%

The hypothesis that 0.9995 would be a beneficial midpoint between 0.999 and 0.9999 is empirically falsified at the 9-epoch budget. The EMA averaging window scales as 1/(1-decay): 0.9995 gives ~2000-step window vs SOTA's ~1000, effectively halving how quickly the EMA tracks model improvements. In our 9-epoch budget where SOTA's val curve is still descending at the final epoch, the slower EMA simply can't catch up.

Key insight (well-articulated by askeladd): "midpoint in decay-space is NOT a midpoint in averaging-window-space." The EMA decay space is now fully closed: 0.998 (bad), 0.999 (SOTA), 0.9995 (bad), 0.9999 (bad). EMA=0.999 confirmed optimal.

W&B run: qvoj4lbs, group: tay-round12-ema-decay-0p9995.
